rllib/core/rl_trainer/scaling_config.py

```diff
@@ -0,0 +1,105 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class RLModuleBackendConfig:
+    """Base class for scaling config relevant to RLTrainer.
+
+    Attributes:
+        distributed: If True, the rl_trainer will be instantiated in distributed mode.
+
+    Methods:
+        set_distributed: Set the distributed flag. _distibuted attribute should not be
+            set to True at the time of constructing the config. The caller should
+            explicitly decide whether the rl_trainer should be instiantiated in
+            distributed mode or not.
+    """
+
+    def __post_init__(self):
+        self._distributed: bool = False
+
+    @property
+    def distributed(self) -> bool:
+        return self._distributed
+
+    def set_distributed(self, distributed: bool) -> "RLModuleBackendConfig":
+        """Set the distributed flag.
+
+        _distibuted attribute should not be set directly at the time of constuction,
+        the caller should explicitly decide whether the rl_trainer should be
+        instiantiated in distributed mode or not.
+
+        Args:
+            distributed: If True, the rl trainer will be instantiated in distributed
+                mode.
+        """
+        self._distributed = distributed
+        return self
+
+
+@dataclass
+class TorchRLModuleBackendConfig(RLModuleBackendConfig):
+    """Torch-specific scaling config relevant to TorchRLTrainer.
+
+    Attributes:
+        use_gpu: If True, the torch rl_trainer will be setup to use the gpu.
+
+    Methods:
+        set_use_gpu: Set the use_gpu flag. _use_gpu attribute should not be set to True
+            at the time of constructing the config. The caller should explicitly decide
+            whether the torch rl_trainer should be using gpu or not.
+    """
+
+    def __post_init__(self):
+        super().__post_init__()
+        self._use_gpu: bool = False
+
+    @property
+    def use_gpu(self) -> bool:
+        return self._use_gpu
+
+    def set_use_gpu(self, use_gpu: bool) -> "TorchRLModuleBackendConfig":
+        """Set the use_gpu flag.
+
+        _use_gpu attribute should not be set directly at the time of constuction,
+        the caller should explicitly decide whether the torch rl_trainer should be
+        using gpu or not
+
+        Args:
+            use_gpu: If True, the rl trainer will be setup to use the gpu.
+        """
+        self._use_gpu = use_gpu
+        return self
+
+
+@dataclass
+class TfRLModuleBackendConfig(RLModuleBackendConfig):
+    """Tf-specific scaling config relevant to TFRLTrainer.
+
+    Args:
+        enable_tf_function: If True, the tf.function decorator will be used to
+            decorate the train_step function. This is recommended to boost performance
+            via tracing the graph.
+    """
+
+    enable_tf_function: bool = True
```

Comment:
Good question: TrainerScalingConfig only contains stuff that are related to instantiating ray actors and each actor's resources. However, that's not enough, the RLTrainer also requires a preparation of its models in a way that will be consistent with the way that the actors are instantiated and assigned resources. For example use_gpu in torch moves the underlying model inside trainer actor onto the correct device. To minimize knowledge leakage going upwards, we need to set those attributes going into the RLTrainer ourselves as a function of the high-level TrainerRunnerConfigs. On the other hand, RLTrainer modules can also be configured via some other parameters upon the request of user. For example enable_eager_tracing provides a trade-off between debuggability and performance. So it depends on the development stage what they want to do. That's why for example enable_eager_tf should be a parameter that is configurabale from outside of trainer runner and gets passed all the way down to the RLTrainer.

Reply:
Just to make sure I understand your response, are we creating a new smaller config message because we don't want to pass the entire config dict to the trainer actors?
also, if you look at the messages, why does Torch get ``use_gpu`` while the TF one gets ``use"_eager_tracing``? why aren't these 2 comparable.
For example, the TF one also should have ``use_gpu``, and the Torch one has ``use_torch_script``.
also, should ScalingConfig be a part of TrainerRunnerConfig?